### PR TITLE
PR #165 and #166, correction in contrib examples

### PR DESCRIPTION
**Pull request by iArpanPatel**
- Example contribution PR ["added Cataract detection #165"](https://github.com/HewlettPackard/swarm-learning/pull/165) with corrections needed to run the example.
- Example contribution PR ["Added nail-disease to examples #166"](https://github.com/HewlettPackard/swarm-learning/pull/166) with corrections needed to run the example.
- Workaround for failure in installing Swarm Learning whl file.
- Existing example contribution documentation correction.

### CODE REVIEW

File: docs/User/Frequently_asked_questions.md [hunk 1]
```diff
@@ -279,6 +279,11 @@ SWOP profile supports mounts with private data. If the installation path or any
 
 No. The program and parameters should be the same across all the Swarm Learning nodes.
 
+### How to resolve "ERROR: Invalid requirement: 'swarmlearning==client': Expected end or semicolon" error while building user container?
+Latest version of pip has become stricter, especially in handling package names and version specifiers, you can resolve this error by downgrading pip to 24.0 in dockerfile or build task yaml file. You can check dockerfile of any of the 'example/contrib/' examples for reference.
+
+By default, when user ML container is run through SWOP or using the `run-sl` script, the user ML container is run with current user's UID and GID of the host machine. If the current user on the host is non-root, the user container also runs as non-root.
```
Comment (suresh-ls) [inline docs/User/Frequently_asked_questions.md:285]: There is extra block of text..   The answer of "How to run user container as non-root?" is repeating here..

File: examples/README.md [hunk 2]
```diff
@@ -24,6 +24,9 @@ Usage: gen-cert -e EXAMPLE-NAME -i HOST-INDEX
 NOTE: HPE recommends that users must use their own certificates in actual production environment.
 </blockquote>
 
+## How to resolve "ERROR: Invalid requirement: 'swarmlearning==client': Expected end or semicolon" error while building user container?
+Latest version of pip has become stricter, especially in handling package names and version specifiers, you can resolve this error by downgrading pip to 24.0 in dockerfile or build task yaml file. You can check dockerfile of any of the 'example/contrib/' examples for reference.
```
Comment (suresh-ls) [inline examples/README.md:28]: Better to split the sentence.      -> "version specifiers. You can resolve this.. "

File: examples/contrib/cataract-detection/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+9.  On both host-1 and host-2, Two node of Swarm training are started. User can monitor the Docker logs of ML nodes \(ML1 and ML2 containers\) for Swarm training on both host-1 and host-2. Training ends with the following log message:
+
+```
+SwarmCallback : INFO : Saved the trained model - model/saved_models/cataract.h5
+```
+
+   Final Swarm model is saved inside the model directory that is `workspace/cataract-detection/model/saved_models` directory on both the hosts. SL and ML nodes exit but it is not removed after the Swarm training.
+
+10. On both host-1 and host-2, To clean up, run the `scripts/bin/stop-swarm` script on all the systems to stop and remove the container nodes of the previous run. If required, backup the container logs and delete the workspace directory.
+
```
Comment (suresh-ls) [inline examples/contrib/cataract-detection/README.md:185]: Better to retain the author references at the end

Comment (suresh-ls) [inline examples/contrib/cataract-detection/README.md:185]: Also better to retain this  ?

"Data-Split:
To reflect a real-life scenario, we have considered different data-splits, where data across different organizations (hospitals/clinics) are unequally distributed containing biased data pertaining to a particular region, sex, ethnicity, or even type of organization. Hence, we have segregated our dataset into three categories such as Gynac (GY) containing a dataset of all women from the age 18-60, General Physician (GP) containing a dataset of men and women up to the age 60 and Senior Citizens (SC) which contains a dataset of all men and women above the age of 60. The maximum validation accuracy obtained for the centralized setup is 93.9, 93.3, and 97.7 for GP, GY, and SC nodes respectively

After training with two swarm learning nodes , the maximum validation accuracy obtained under the Swarm Learning setup is 95.48 and 98.27 for GP-GY and GP-SC splits respectively

The following summarizes our data splits
1) GP
2) GY
3) SC"

File: examples/contrib/nail-disease/README.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+It also uses pre-trained model(InceptionV3) to leverage the concept of transfer learning
+
+The Swarm training of pre-trained InceptionV3 model using two ML nodes is demonstrated in this example. The ML environment is built manually using docker command and Swarm training is launched without SWCI or SWOP nodes. Here `scripts/bin/run-sl` script is used to spawn each ML node to run InceptionV3 model as a `sidecar` container of each SL node.
+
+This project uses dataset from multiple sources :
+
+* https://www.kaggle.com/datasets/arthitaya/nail-dataset2
+
+* https://www.kaggle.com/datasets/arthitaya/nail-dataset
+
```
Comment (suresh-ls) [inline examples/contrib/nail-disease/README.md:13]: Why is this dataset reference dropped ? 
* https://www.kaggle.com/datasets/reubenindustrustech/nail-dataset

File: examples/contrib/cataract-detection/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+# Privacy preserving Swarm Learning on Cataract Detection
+
+Cataract is one of the most common issues faced during old age 
+that in the worst case can lead to loss of sight. Detecting 
+cataract during its early stages can prevent several complications.
+Data in the health care industry is considered to be highly sensitive and is distributed in various hospitals. Hence this project, focuses on developing a machine learning framework using Swarm Learning to help in the early detection of cataracts by maintaining the privacy of the patients data while creating an accurate and a robust model.
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/README.md:7]: health care -> remove between space

File: examples/contrib/cataract-detection/README.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+It contains our model integrated with Swarm Learning Framework.
+
+This project shows the Swarm training of Keras based Machine Learning 
+model using two ML nodes. It also shows how ML environment can 
+be built manually using docker command and how Swarm training 
+can be launched without SWCI or SWOP nodes. 
+Here `scripts/bin/run-sl` script is used to spawn each ML node 
+to run Cataract model as a `sidecar` container of each SL node.
+
+The following image illustrates a cluster setup:
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/README.md:53]: No image reference here. High level was on top.

File: examples/contrib/cataract-detection/README.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+
+-   This example uses one SN node. SN1 is the name of the Docker container that runs on host 172.1.1.4.
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/README.md:56]: is there any need to start with .4 and .5 hosts? why can't we use .1 and .2?

File: examples/contrib/cataract-detection/ml-context/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+keras
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/ml-context/requirements.txt:1]: Suggestion - Fix version numbers for these packages also, so that examples can run in future if these packages have breaking changes

File: examples/contrib/cataract-detection/model/cataract.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    max_epochs = int(os.getenv('MAX_EPOCHS', str(default_max_epochs)))
+    min_peers = int(os.getenv('MIN_PEERS', str(default_min_peers)))
+
+    (x_train, y_train), (x_test, y_test) = load_data(dataDir)
+
+    save_dir = os.path.join(modelDir, 'saved_models')
+    saved_model_name = 'cataract.h5'
+
+    image_size = 224
+    vgg = VGG19(weights="imagenet", include_top=False,input_shape=(image_size, image_size, 3))
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/model/cataract.py:39]: We can add sentence in readme to state this example also uses transfer learning.

File: examples/contrib/cataract-detection/model/cataract.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    model.add(vgg)
+    model.add(Flatten())
+    model.add(Dense(1, activation="sigmoid"))
+    print(model.summary())
+
+
+    model.compile(optimizer="adam", loss="binary_crossentropy",
+                  metrics=["accuracy"])
+
+    swarmCallback = SwarmCallback(syncFrequency=128, minPeers=min_peers,useAdaptiveSync=False, adsValData=(x_test, y_test), adsValBatchSize=8)
```
Comment (RadhakrishnaJ) [inline examples/contrib/cataract-detection/model/cataract.py:54]: SwarCallback parameters are important tunable parameters so for better readability have them in line by line.

File: examples/contrib/nail-disease/model/nail-disease.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+
+swarmCallback.logger.info('Saving the final Swarm model ...')
+if not os.path.isdir(save_dir):
+    os.makedirs(save_dir)
+model_path = os.path.join(save_dir, model_name)
+model.save(model_path)
+swarmCallback.logger.info(f'Saved the trained model - {model_path}')
+
+loss, acc = model.evaluate(validation_generator)
```
Comment (RadhakrishnaJ) [inline examples/contrib/nail-disease/model/nail-disease.py:109]: Please add Test document to show commands works as is and test produces results as documented.

File: docs/User/Frequently_asked_questions.md (hunk 1 above)
Comment (iArpanPatel) [reply docs/User/Frequently_asked_questions.md:285]: cleaned

File: examples/README.md (hunk 2 above)
Comment (iArpanPatel) [reply examples/README.md:28]: Done

File: examples/contrib/cataract-detection/README.md (hunk 5 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/README.md:7]: done

File: examples/contrib/cataract-detection/README.md (hunk 6 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/README.md:53]: Corrected

File: examples/contrib/cataract-detection/README.md (hunk 3 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/README.md:185]: retained

Comment (iArpanPatel) [reply examples/contrib/cataract-detection/README.md:185]: restored

File: examples/contrib/nail-disease/README.md (hunk 4 above)
Comment (iArpanPatel) [reply examples/contrib/nail-disease/README.md:13]: corrected

File: examples/contrib/cataract-detection/README.md (hunk 7 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/README.md:76]: corrected.

File: examples/contrib/cataract-detection/model/cataract.py (hunk 9 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/model/cataract.py:39]: done.

File: examples/contrib/cataract-detection/ml-context/requirements.txt (hunk 8 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/ml-context/requirements.txt:1]: This needs to be fixed in all other existing examples, due time limitation for now we will defer this to next phase of updates.

**Review by suresh-ls — APPROVED** [review]
Looks fine.. As discussed, please upload the testing done & logs to this PR before closing this PR.   Thanks!

**Comment by iArpanPatel** [comment]
[cataract-sl1.txt](https://github.com/user-attachments/files/17568505/cataract-sl1.txt)
[cataract-sl2.txt](https://github.com/user-attachments/files/17568506/cataract-sl2.txt)
[cataract-ml1.txt](https://github.com/user-attachments/files/17568508/cataract-ml1.txt)
[cataract-ml2.txt](https://github.com/user-attachments/files/17568511/cataract-ml2.txt)
[nail-sl1.txt](https://github.com/user-attachments/files/17568514/nail-sl1.txt)
[nail-sl2.txt](https://github.com/user-attachments/files/17568515/nail-sl2.txt)
[nail-ml1.txt](https://github.com/user-attachments/files/17568516/nail-ml1.txt)
[nail-ml2.txt](https://github.com/user-attachments/files/17568517/nail-ml2.txt)

File: examples/contrib/nail-disease/model/nail-disease.py (hunk 11 above)
Comment (iArpanPatel) [reply examples/contrib/nail-disease/model/nail-disease.py:109]: Added here.

File: examples/contrib/cataract-detection/model/cataract.py (hunk 10 above)
Comment (iArpanPatel) [reply examples/contrib/cataract-detection/model/cataract.py]: done

**Review by RadhakrishnaJ — APPROVED** [review]
